### PR TITLE
Support for dot(dns, csr) = dns and dot(dns, csr.T) = dns on CPU

### PR DESCRIPTION
**Pull request by XiaotaoChen**
Description

Support for dot(dns, csr) = dns and dot(dns, csr.T) = dns on CPU

Checklist

Essentials

Please feel free to remove inapplicable items for your PR.

- The PR title starts with [MXNET-JIRA_ID], where JIRA_ID refers to the relevant JIRA issue created (except PRs with tiny changes)
- Changes are complete (i.e. I finished coding on this PR)
- All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-PR_ID/BUILD_ID/index.html
- To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

Changes

- Feature1, tests, (and when applicable, API doc)
- Feature2, tests, (and when applicable, API doc)

Comments

We implemented  dot(dns,csr/csr.T)=dns on cpu inspired by the implementation  of dot(dns,csr/csr.T)=dns on gpu by @haojin2 . It gains 190 times speed up when density of csr is 0.01%. The details as follows.

    ---test dot(dns1, csr) = dns2, shape_lhs=(256,30000), shape_rhs=(30000,30000)---
    1.00 % with fallback: 2.205663, without fallback: 1.179663
    0.50 % with fallback: 4.471481, without fallback: 2.309979
    0.10 % with fallback: 20.552546, without fallback: 10.095501
    0.05 % with fallback: 39.141270, without fallback: 18.930919
    0.01 % with fallback: 155.199133, without fallback: 75.759737
    ---test dot(dns1, csr.T) = dns2, shape_lhs=(256,30000), shape_rhs=(30000,30000)---
    1.00 % with fallback: 3.031672, without fallback: 1.474902
    0.50 % with fallback: 6.197231, without fallback: 3.054169
    0.10 % with fallback: 29.555133, without fallback: 14.591607
    0.05 % with fallback: 56.621382, without fallback: 28.377709
    0.01 % with fallback: 190.064799, without fallback: 94.031036

The benchmark script is here: https://github.com/XiaotaoChen/incubator-mxnet/blob/mytest/example/sparse/cxt-test/test_dot.py . And also can test this with haojin's script by replacing mx.gpu() with mx.cpu() : https://github.com/apache/incubator-mxnet/pull/10371 
@pengzhao-intel @TaoLv

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
@haojin2 @eric-haibin-lin for the review :)

File: src/operator/tensor/dot-inl.h [hunk 1]
```diff
@@ -264,11 +264,15 @@ inline bool DotForwardInferStorageType(const nnvm::NodeAttrs& attrs,
...
-    // dns, csr -> csr on CPU
-    if (dev_mask == mshadow::cpu::kDevMask && !param.transpose_b) {
-      if (target_stype == kCSRStorage) {
+    if (dev_mask == mshadow::cpu::kDevMask) {
+      // dns, csr -> csr on CPU
+      if (target_stype == kCSRStorage && !param.transpose_b) {
         dispatched = storage_type_assign(&out_stype, kCSRStorage, dispatch_mode,
                                          DispatchMode::kFComputeEx);
+      // dns, csr/csr.T -> dns on CPU
+      } else if (target_stype ==  kDefaultStorage) {
```
Comment (haojin2) [inline src/operator/tensor/dot-inl.h]: nit: extra space at "==  kDefaultStorage"

File: src/operator/tensor/dot-inl.h [hunk 2]
```diff
@@ -327,7 +331,8 @@ inline bool DotBackwardInferStorageType(const nnvm::NodeAttrs& attrs,
       dispatched = true;
     }
   }
-  if (!dispatched && dev_mask == mshadow::gpu::kDevMask && !param.transpose_a &&
+  // if (!dispatched && dev_mask == mshadow::gpu::kDevMask && !param.transpose_a &&
```
Comment (haojin2) [inline src/operator/tensor/dot-inl.h]: Please remove unused code.

File: src/operator/tensor/dot-inl.h [hunk 3]
```diff
@@ -1031,13 +1110,58 @@ inline void DotDnsCsrCsrImpl(const OpContext& ctx, const cpu& cpu_dev,
...
+        num_threads = mxnet_op::get_num_threads<cpu>(data_out.shape_[0]);
+        // seg by output row
+        dim_t seg_len = (data_out.shape_[0] + num_threads - 1) / num_threads;
+        if (transpose_b) {
+          mxnet_op::Kernel<DotDnsCsrTransDnsByRowBlocks, cpu>::Launch(s, num_threads,
+              data_out.dptr<DType>(), data_l.dptr<DType>()
+              , data_r.dptr<DType>(), indptr_r.dptr<IType>()
+              , col_idx_r.dptr<CType>(), seg_len
+              , dns.shape_[0], dns.shape_[1]
+              , rhs.shape()[0], rhs.shape()[1]);
```
Comment (haojin2) [inline src/operator/tensor/dot-inl.h]: For line breaks please follow google c++ style guide, a sample may be found here: https://gist.github.com/davidzchen/9187878#file-sample-google-c-L110-L114

File: src/operator/tensor/dot-inl.h [hunk 4]
```diff
@@ -1031,13 +1110,58 @@ inline void DotDnsCsrCsrImpl(const OpContext& ctx, const cpu& cpu_dev,
...
 /*
- * \brief Impl of dot(dns, csr) = dense (GPU only)
+ * \brief Impl of dot(dns, csr) = dns and dot(dns, csr.T) = dns
  */
 inline void DotDnsCsrDnsImpl(const OpContext& ctx, const cpu& cpu_dev,
                              const TBlob& dns, const NDArray& rhs,
                              const OpReqType req, NDArray* ret,
                              const bool transpose_b) {
-  LOG(FATAL) << "dot(dense, csr) = dense is not implemented on CPU";
+  if (kNullOp == req) return;
```
Comment (haojin2) [inline src/operator/tensor/dot-inl.h]: nit: personally I would prefer "req == kNullOp".

File: src/operator/tensor/dot-inl.h [hunk 5]
```diff
@@ -1031,13 +1110,58 @@ inline void DotDnsCsrCsrImpl(const OpContext& ctx, const cpu& cpu_dev,
...
+  const TBlob indptr_r = rhs.aux_data(csr::kIndPtr);
+  const TBlob col_idx_r = rhs.aux_data(csr::kIdx);
+  const TBlob& data_l = dns;
+  const TBlob data_out = ret->data();
+
+  MSHADOW_SGL_DBL_TYPE_SWITCH(data_r.type_flag_, DType, {  // data type
+    MSHADOW_IDX_TYPE_SWITCH(indptr_r.type_flag_, IType, {  // indptr type
+      MSHADOW_IDX_TYPE_SWITCH(col_idx_r.type_flag_, CType, {  // col idx type
+        dim_t num_threads;
+        if (kWriteTo == req) {
```
Comment (haojin2) [inline src/operator/tensor/dot-inl.h]: nit: Personally I would prefer "req == kWriteTo"

Comment (XiaotaoChen) [reply src/operator/tensor/dot-inl.h]: Thx, I have modified those according to your suggestion.

File: src/operator/tensor/dot-inl.h [hunk 6]
```diff
@@ -264,11 +264,15 @@ inline bool DotForwardInferStorageType(const nnvm::NodeAttrs& attrs,
   if (!dispatched && lhs_stype == kDefaultStorage && rhs_stype == kCSRStorage &&
       !param.transpose_a) {
     target_stype = hint_has_value ? target_stype : kCSRStorage;
-    // dns, csr -> csr on CPU
```
Comment (eric-haibin-lin) [inline src/operator/tensor/dot-inl.h]: Please also update doc in https://github.com/apache/incubator-mxnet/blob/master/src/operator/tensor/dot.cc#L63-L64

File: src/operator/tensor/dot-inl.h [hunk 7]
```diff
@@ -655,7 +659,81 @@ struct DotDnsCsrCsrByRowBlocks {
...
+};
 
+/*!
+ * \brief CPU Kernel of dot(dns1, csr.T) = dns2
+ * Parallelization by row blocks
+ */
+struct DotDnsCsrTransDnsByRowBlocks {
+  /*!
+   * \brief
+   * \param i the i-th thread
```
Comment (eric-haibin-lin) [inline src/operator/tensor/dot-inl.h]: Please complete documentation

**Comment by XiaotaoChen** [comment]
@haojin2 @eric-haibin-lin  I have solved the problems mentioned in your comments. Can you accept this PR ?

**Comment by haojin2** [comment]
LGTM, will wait for @eric-haibin-lin to take a final look.

File: src/operator/tensor/dot-inl.h [hunk 8]
```diff
@@ -1031,13 +1129,58 @@ inline void DotDnsCsrCsrImpl(const OpContext& ctx, const cpu& cpu_dev,
...
+
+  using nnvm::dim_t;
+
+  const TBlob data_r = rhs.data();
+  const TBlob indptr_r = rhs.aux_data(csr::kIndPtr);
+  const TBlob col_idx_r = rhs.aux_data(csr::kIdx);
+  const TBlob& data_l = dns;
+  const TBlob data_out = ret->data();
+
+  MSHADOW_SGL_DBL_TYPE_SWITCH(data_r.type_flag_, DType, {  // data type
```
Comment (eric-haibin-lin) [inline src/operator/tensor/dot-inl.h]: fp16 will fail with this branch. I think there's a MSHADOW_REAL_TYPE_SWITCH

File: src/operator/tensor/dot-inl.h [hunk 9]
```diff
@@ -1031,13 +1129,58 @@ inline void DotDnsCsrCsrImpl(const OpContext& ctx, const cpu& cpu_dev,
...
+    MSHADOW_IDX_TYPE_SWITCH(indptr_r.type_flag_, IType, {  // indptr type
+      MSHADOW_IDX_TYPE_SWITCH(col_idx_r.type_flag_, CType, {  // col idx type
+        dim_t num_threads;
+        if (req == kWriteTo) {
```
Comment (eric-haibin-lin) [inline src/operator/tensor/dot-inl.h]: req == writeto || req == writeinplace

Comment (XiaotaoChen) [reply src/operator/tensor/dot-inl.h]: Thx, it's Done ! May you review it again :)

**Comment by XiaotaoChen** [comment]
Thx, it's Done ! May you review it again :) @eric-haibin-lin

**Pull request merged**